keps/sig-cli/kustomize-pod-spec-patch.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Kustomize PodSpec Patch
+authors:
+ - "@pwittrock"
+owning-sig: sig-cli
+participating-sigs:
+ - sig-apps
+reviewers:
+ - "@Liujingfang1"
+ - "@janetkuo"
+approvers:
+ - "@monopole"
+editors:
+ - "@pwittrock"
+creation-date: 2019-03-01
+last-updated: 2019-03-01
+status: provisional
+see-also:
+replaces:
+superseded-by:
+ - n/a
+---
+
+# Kustomize PodSpec Patch
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+A common need for users is to inject sidecar containers across many different workloads.
+While overlays and patches allow injection on a per-Resource basis, they don't
+provide a mechanism for injecting the same patch across multiple workloads.
+
+## Motivation
+
+Cross-cutting PodSpec customizations are necessary for usecases such as:
+
+- Injecting Envoy proxies into all Workloads
+- Injecting log rotation containers into all Workloads
+- Setting the ServiceAccount for Workloads matching a query
+
+### Goals
+
+Allow users to define cross-cutting PodSpec customizations by matching the Resource annotations.
+
+
+### Non-Goals
+
+## Proposal
+
+Add field to `kustomization.yaml`: `podSpecPatches`.  It has both a file with a `PodSpec`
+that will be patched, and a `annotationSelector` defining the Resources that will be patched.
+
+Resources matching the annotationSelector that have a PodSpec in their schema will be
+patched.
+
+```go
+
+type Kustomization struct {
+	// ...
+	
+	patchesPodSpec []PatchPodSpec `json:"patchesPodSpec"`
+	// ...
+}
+
+// ResourceSelector selects Resources that it matches
+type PatchPodSpec struct {
```

Comment:
So `patchesPodSpec` is a kind of patch that can be applied to multiple targets. In `patchesStrategicMerge` and `patchesJson6902`, kustomize uses group, version, kind and name to find the target resource to patch. I think we can also add group, version, kind and name to `patchesPodSpec`. Those fields are all optional and the name could support regular expression.

Reply:
What is the use case for specifying those?

Reply:
For example, applying the patch to all Deployments or applying the patch to all resources with names "app1-*".